experimental/proxy_attestation/src/main.rs

```diff
@@ -0,0 +1,168 @@
+//! - Proxy sends the client its root certificate
+//!     - Client trusts that this is a correct certificate, since it was sent via a secure channel
+//!       created during the attestation process
+//! - Client connects to the application using TLS, and the application uses a TLS certificate
+//!   previously signed by the proxy
+//! - Client checks that the certificate was signed by the root certificate and establishes a secure
+//!   connection
+//!
+//! Note: Current version of Remote Attestation Service doesn't have remote attestation.
+
+mod certificate;
+
+use crate::certificate::CertificateAuthority;
+use anyhow::Context;
+use futures::future::FutureExt;
+use log::info;
+use oak_proxy_attestation::proto::{
+    proxy_attestation_server::{ProxyAttestation, ProxyAttestationServer},
+    GetRootCertificateRequest, GetRootCertificateResponse, GetSignedCertificateRequest,
+    GetSignedCertificateResponse,
+};
+use openssl::x509::X509Req;
+use structopt::StructOpt;
+use tonic::{
+    transport::{Identity, Server, ServerTlsConfig},
+    Request, Response, Status,
+};
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Aggregator Backend")]
+pub struct Opt {
+    #[structopt(long, help = "Private RSA key file used by gRPC server.")]
+    grpc_tls_private_key: String,
+    #[structopt(
+        long,
+        help = "PEM encoded X.509 TLS certificate file used by gRPC server."
+    )]
+    grpc_tls_certificate: String,
+}
+
+pub struct Proxy {
+    certificate_authority: CertificateAuthority,
+}
+
+impl Proxy {
+    fn new() -> Self {
+        Self {
+            certificate_authority: CertificateAuthority::new(),
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl ProxyAttestation for Proxy {
+    /// Creates a signed X.509 certificate based on the certificate signing request.
+    async fn get_signed_certificate(
+        &self,
+        req: Request<GetSignedCertificateRequest>,
+    ) -> Result<Response<GetSignedCertificateResponse>, Status> {
+        info!("Received certificate sign request");
+        let certificate_request = X509Req::from_pem(&req.into_inner().certificate_request)
+            .map_err(|error| {
+                Status::invalid_argument(format!(
+                    "Certificate deserialize certificate request: {}",
+                    error
+                ))
+            })?;
+        match self
+            .certificate_authority
+            .sign_certificate(certificate_request)
+        {
+            Ok(certificate) => match certificate.to_pem() {
+                Ok(serialized_certificate) => {
+                    info!("Sending signed certificate");
+                    let mut res = GetSignedCertificateResponse::default();
+                    res.certificate = serialized_certificate;
+                    Ok(Response::new(res))
+                }
+                Err(error) => Err(Status::internal(format!(
+                    "Certificate serialization error: {}",
+                    error
```

Comment:
Please don't just send unsanitized error messages to clients like this, they may contain all sort of internal data. Also, it seems that the server is not even logging these errors. I think exactly the opposite should happen: log on the server, and send a generic message back to the client. Otherwise it is basically impossible to debug.

Reply:
Done